File: tokenmaxxing-macos/Sources/Tokenmaxxing/Format.swift
import Foundation

/// Compact number formatting shared across the dashboard and export.
enum Fmt {
    static func count(_ n: Int64) -> String {
        let value = Double(n)
        if value >= 1e12 { return String(format: "%.1fT", value / 1e12) }
        if value >= 1e9 { return String(format: "%.1fB", value / 1e9) }
        if value >= 1e6 { return String(format: "%.1fM", value / 1e6) }
        if value >= 1e3 { return String(format: "%.1fK", value / 1e3) }
        return "\(n)"
    }

    static func usd(_ v: Double) -> String {
        let a = abs(v)
        if a >= 1e6 { return String(format: "$%.2fM", v / 1e6) }
        if a >= 10_000 { return String(format: "$%.1fK", v / 1e3) }
        if a >= 100 { return String(format: "$%.0f", v) }
        return String(format: "$%.2f", v)
    }

    static func usdCents(_ v: Double) -> String { String(format: "$%.2f", v) }

    static func until(_ seconds: Int) -> String {
        let s = max(0, seconds)
        let (days, hours, minutes) = (s / 86_400, (s % 86_400) / 3_600, (s % 3_600) / 60)
        if days > 0 { return "\(days)d \(hours)h" }
        if hours > 0 { return "\(hours)h \(minutes)m" }
        return "\(minutes)m"
    }

    static func until(_ date: Date) -> String {
        until(Int(date.timeIntervalSinceNow))
    }

    static func percent(_ fraction: Double) -> String {
        "\(Int((fraction * 100).rounded()))%"
    }
}

File: tokenmaxxing-kde/src/format.rs
//! Compact number formatting shared across the dashboard and share card.

/// A count as a short human string: `1234` → `1.2K`, `3_400_000` → `3.4M`.
pub fn count(n: u64) -> String {
    let value = n as f64;
    if value >= 1e12 {
        format!("{:.1}T", value / 1e12)
    } else if value >= 1e9 {
        format!("{:.1}B", value / 1e9)
    } else if value >= 1e6 {
        format!("{:.1}M", value / 1e6)
    } else if value >= 1e3 {
        format!("{:.1}K", value / 1e3)
    } else {
        n.to_string()
    }
}

/// Dollars, compacted for headline tiles: `$12.3K`, `$1.2M`, else two decimals.
pub fn usd(v: f64) -> String {
    let a = v.abs();
    if a >= 1e6 {
        format!("${:.2}M", v / 1e6)
    } else if a >= 10_000.0 {
        format!("${:.1}K", v / 1e3)
    } else if a >= 100.0 {
        format!("${:.0}", v)
    } else {
        format!("${v:.2}")
    }
}

/// Dollars always to the cent — for spend caps and precise figures.
pub fn usd_cents(v: f64) -> String {
    format!("${v:.2}")
}

/// A duration until `seconds` from now, as `3d 4h` / `5h 12m` / `9m`.
pub fn until(seconds: i64) -> String {
    let s = seconds.max(0);
    let (days, hours, minutes) = (s / 86_400, (s % 86_400) / 3_600, (s % 3_600) / 60);
    if days > 0 {
        format!("{days}d {hours}h")
    } else if hours > 0 {
        format!("{hours}h {minutes}m")
    } else {
        format!("{minutes}m")
    }
}

/// A percentage as `73%`.
pub fn percent(fraction: f64) -> String {
    format!("{}%", (fraction * 100.0).round() as i64)
}
